source/common/config/filesystem_subscription_impl.h

```diff
@@ -33,10 +33,17 @@ class FilesystemSubscriptionImpl : public Config::Subscription<ResourceType>,
     // We report all discovered resources in the watched file.
     UNREFERENCED_PARAMETER(resources);
     callbacks_ = &callbacks;
-    watcher_->addWatch(path_, Filesystem::Watcher::Events::MovedTo, [this](uint32_t events) {
-      UNREFERENCED_PARAMETER(events);
-      refresh();
-    });
+    try {
+      watcher_->addWatch(path_, Filesystem::Watcher::Events::MovedTo, [this](uint32_t events) {
+        UNREFERENCED_PARAMETER(events);
+        refresh();
+      });
+    } catch (const EnvoyException& e) {
+      ENVOY_LOG(warn,
```

Comment:
Shouldn't this exception propagate and then cause the config to fail to load or the server to shutdown? I guess I'm just trying to understand if this kind of error can reasonably happen and if we should keep running?

Reply:
I think there are two general failure cases:

1) Can't access the path to watch (doesn't exist, permissions, etc)
2) Out of file-descriptors (or other ulimit)

For (1), I think the error should propagate up so we know we have a bad config.
For (2), if we want to try to detect it separately from 1, we should treat it the same as other out-of-fd errors, which I think envoy abort()'s on. But I'd be ok with lumping this failure in with (1). It'll be nasty trying to sort out which type of failure it is on all platforms (we currently have different kqueue and inotify implementations).

Reply:
I've done some digging into this. The issue is that we don't see these errors at config load time, since we only add the watch when we `start()` the subscription, which occurs later, for example at bootstrap ingest when we are at https://github.com/envoyproxy/envoy/blob/e26898e045d724cf101044ce8c2a18c838ec9798/source/server/server.cc#L360.

I don't think we do a great job at handling runtime exceptions in general, as this has exposed. I think there are two concrete distinct things to do here:

1. Move the `addWatch()` to the constructor for `FilesystemSubscriptionImplTest()`. We should see the failure appear at config load. We can ignore any updates until `start()` is invoked.

2. Add better behavior for runtime exceptions. I'm not sure where we want to go here; if we hit one of these, something Bad has happened, we've totally violated some internal invariant. Do we want graceful handling here, or do we want a segfault and bug report?

Re: @ggreenway I agree, these are distinct, but for now I'll just lump (2) with (1), otherwise we're playing interpretive games with errnos.

Reply:
@htuch there might be other open issues on this, but in general I think we are basically totally FUBARd with how we deal with files for xDS in the runtime scenario. E.g., someone deletes the file while envoy is running, etc. I'm pretty sure we will break badly in many cases. I think @junr03 added some TODOs in this area. I'm not sure if there is an issue tracking, but we should open one. I honestly would not feel comfortable running with file based xDS in production right now, so we should probably have some big warning somewhere in the docs on this...

Obviously runtime exceptions should be handled properly and gracefully where possible. Otherwise I think they should crash us as was already done.

Reply:
I think my preference for now, for the general "unexpected exception was thrown" case is to let it crash and get a bug report. For this specific case, I agree that this code should get moved to config-load time so we can handle the error appropriately.